src/CoreFoundation/CFUrl.cs

```diff
@@ -45,62 +47,32 @@ public enum CFUrlPathStyle : long {
 	};
 
 	// CFURL.h
-	public class CFUrl
-#if !COREBUILD
-		: INativeObject, IDisposable
-#endif
+	public class CFUrl : NativeObject
 	{
 #if !COREBUILD
-		internal IntPtr handle;
-
-		~CFUrl ()
-		{
-			Dispose (false);
-		}
-		
-		public void Dispose ()
-		{
-			Dispose (true);
-			GC.SuppressFinalize (this);
-		}
-
-		public IntPtr Handle {
-			get { return handle; }
-		}
-		
-		protected virtual void Dispose (bool disposing)
-		{
-			if (handle != IntPtr.Zero){
-				CFObject.CFRelease (handle);
-				handle = IntPtr.Zero;
-			}
-		}
-
 		[DllImport (Constants.CoreFoundationLibrary)]
 		extern static /* CFURLRef */ IntPtr CFURLCreateWithFileSystemPath (/* CFAllocatorRef */ IntPtr allocator, 
 			/* CFStringRef */ IntPtr filePath, 
 			/* CFURLPathStyle */ nint pathStyle, 
 			/* Boolean */ [MarshalAs (UnmanagedType.I1)] bool isDirectory);
 		
-		internal CFUrl (IntPtr handle)
+		internal CFUrl (IntPtr handle, bool owns)
+			: base (handle, owns)
 		{
-			this.handle = handle;
 		}
 		
-		internal CFUrl (IntPtr handle, bool owned)
+		static public CFUrl? FromFile (string filename)
 		{
-			if (!owned)
-				CFObject.CFRetain (handle);
-			this.handle = handle;
-		}
-		
-		static public CFUrl FromFile (string filename)
-		{
-			using (var str = new CFString (filename)){
-				IntPtr handle = CFURLCreateWithFileSystemPath (IntPtr.Zero, str.Handle, (nint)(long)CFUrlPathStyle.POSIX, false);
+			if (filename is null)
+				throw new ArgumentNullException (nameof (filename));
+			var strHandle = CFString.CreateNative (filename);
+			try {
+				var handle = CFURLCreateWithFileSystemPath (IntPtr.Zero, strHandle, (nint)(long)CFUrlPathStyle.POSIX, false);
 				if (handle == IntPtr.Zero)
 					return null;
 				return new CFUrl (handle, true);
+			} finally {
+				CFString.ReleaseNative (strHandle);
```

Comment:
Just a thought, a `ref readonly struct` with `Dispose()` defined can be used to simplify the code. Would probably need a disposer struct for most common ones like CFString, but code does look a bit nicer.

https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEmuAdAEoCuMCUYAplQMID2YADlADYsATgGVhANygBjFigDcRTAGYK2MuzIBvImV0UVmACxkAsrgAUASi069d8RCFkeCJwF4yAcRYIAEhBgAE0ErBUI7O1cATxtwiPiAEU4RBDoAMzTzFyFLMPjdAF8yNNgIfn4Y7Tj8vXYAMRoWQVQWLNdc23iCzrJu6rIe5QpjE2wrHqqa0jIHJ2yyDxgWAHcyBKgUHk4UYXNvPwDg1ssO/rsklPTM7KoAfWzTuz67QYMRpXH+yfzZ51cFrw+fxBEKPfLTX63AFLVbrTbbXYPPL5C6pDJtHLIwo9HpCFgQQKcGAVMh4tJkFCuOhSBBrDZbHZCCY9OxDPEEokkgCSjAACv97q4sREhnCGbseQh+XN2iy9N8arpBe4/kJhU85bpNfphnT4TtPjUFYr6o1mgblWCuj1nvkegBtEw+AAWnECXN4/HMToQrvdnoA8i4oESUFQAHKcHn8WCwADmlgAuoNcAA2XWmpr4g2S6WqyzM/p9B0+v0enhe0tu8v8INMUMRqPE2MwBPJ/qUdO5/77YFHQ3y7WYADsZG7QioAC1hJxhcX+o6XdXPd6l/6K3WQzAw5Hoy22yn00Y1sk0Zlx/nCxrCEQCkA==

Reply:
It looks even cleaner now with some other changes, M4 is pretty clean looking and generates identical code.

[sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEmuAdAEoCuMCUYAplQMID2YADlADYsATgGVhANygBjFigDcRTAGYKuAGwVsZAKIAPBEQDeRMqYorKG9gDEaLQamFkAIiwBmwm3YcoWACgQACygUMgBJRgAFBCEyHhiAShMzY0IzdIoAdjIYFgB3Mi97FkchP3ihBIU0swBfZNMG8zIhEoATThh+AE8W9zIUGLopBELbYtKm1IyzZT6IDq7eiIRo2IB9CuqZ0zmin2E/FbW4xKb06Z30zZiyAF5ToW2d+pqZ89mLABYXEJ5OXx+JJvDKXK6mfYlQE3SrPGavK4fUwAemRZAAgvx+JwCpDSmQEJwyMAWHFUL42mRYISxt4oSwkc1LFTePxpFBRpweMIIITYscYn48b5Yq0DpVGWDwZhsmL6UIqDC4RkEXVGU0ANoAWRYQU4bTCrL8Or1BtZAHl4lBOigqAA5TgRNkwWAAcwSAF0mnNmZgfsL/ALYhVgRcmq9XooLFp2GQpT6flrcEC44zxBBg7cHgBxXUACQgMDagiByvSMV6Uquzk4IgQdDcbnKiTLdTIblgECxlcZ6T0CA44wOzdhjNVpnHZG93zIWuwQKmjNIZHTmdiD1yBX7g7ppT8uYQBaLJYSVUZNbrDabFUVIeVk+nFETSgXIKrZlXj3uZAPR+L/jPEF0mXT91m/TcdH0HcJhFEdAOrWt60bOD7yaR8/VnL5XzDIDZnIT8Km/X9C3/IEqFcDwhADUtz0Qq8UPDIhNRNQJ9UNHh+GNXVWLNDjLSYG17UdLpYDdT1vXUcIoluYjjwAxdcN2bIgyoAAtYROFQwhmO4tijRYvS+KtQSHSdUSYHdL0QV9H4LyQpsg0eUMUkY7TCFqIA===)